cpp/src/components/weak_cc.cuh

```diff
@@ -0,0 +1,343 @@
+                if(ci<cj) {
+                    atomicMin(labels + row_ind_ptr[start +j], ci);
+                    xa[row_ind_ptr[start+j]] = true;
+                    m[0] = true;
+                }
+                else if(ci>cj) {
+                    ci = cj;
+                    ci_mod = true;
+                }
+            }
+            if(ci_mod) {
+                atomicMin(labels + startVertexId + tid, ci);
+                xa[startVertexId + tid] = true;
+                m[0] = true;
+            }
+        }
+    }
+}
+
+
+template <typename Type, int TPB_X = 32, typename Lambda>
+__global__ void weak_cc_init_label_kernel(Type *labels, Type startVertexId, Type batchSize,
+        Type MAX_LABEL, Lambda filter_op) {
+    /** F1 and F2 in the paper correspond to fa and xa */
+    /** Cd in paper corresponds to db_cluster */
+    Type tid = threadIdx.x + blockIdx.x*TPB_X;
+    if(tid<batchSize) {
+        if(filter_op(tid) && labels[tid + startVertexId]==MAX_LABEL)
+            labels[startVertexId + tid] = Type(startVertexId + tid + 1);
+    }
+}
+
+template <typename Type, int TPB_X = 32>
+__global__ void weak_cc_init_all_kernel(Type *labels, bool *fa, bool *xa,
+        Type N, Type MAX_LABEL) {
+    Type tid = threadIdx.x + blockIdx.x*TPB_X;
+    if(tid<N) {
+        labels[tid] = MAX_LABEL;
+        fa[tid] = true;
+        xa[tid] = false;
+    }
+}
+
+  template <typename Type, int TPB_X = 32, typename Lambda>
+void weak_cc_label_batched(Type *labels,
+        const Type* row_ind, const Type* row_ind_ptr, Type nnz, Type N,
+        WeakCCState<Type> *state,
+        Type startVertexId, Type batchSize,
+        cudaStream_t stream, Lambda filter_op) {
+    bool host_m;
+    bool *host_fa = (bool*)malloc(sizeof(bool)*N);
+    bool *host_xa = (bool*)malloc(sizeof(bool)*N);
+
+    dim3 blocks(ceildiv(batchSize, TPB_X));
+    dim3 threads(TPB_X);
+    Type MAX_LABEL = std::numeric_limits<Type>::max();
+
+    weak_cc_init_label_kernel<Type, TPB_X><<<blocks, threads, 0, stream>>>(labels,
+            startVertexId, batchSize, MAX_LABEL, filter_op);
+    CUDA_CHECK(cudaPeekAtLastError());
+    do {
+        CUDA_CHECK( cudaMemsetAsync(state->m, false, sizeof(bool), stream) );
+        weak_cc_label_device<Type, TPB_X><<<blocks, threads, 0, stream>>>(
+                labels,
+                row_ind, row_ind_ptr, nnz,
+                state->fa, state->xa, state->m,
+                startVertexId, batchSize);
+        CUDA_CHECK(cudaPeekAtLastError());
+
+        //** swapping F1 and F2
+        MLCommon::updateHost(host_fa, state->fa, N, stream);
+        MLCommon::updateHost(host_xa, state->xa, N, stream);
+        MLCommon::updateDevice(state->fa, host_xa, N, stream);
+        MLCommon::updateDevice(state->xa, host_fa, N, stream);
+
+        //** Updating m *
+        MLCommon::updateHost(&host_m, state->m, 1, stream);
+        CUDA_CHECK(cudaStreamSynchronize(stream));
+    } while(host_m);
+}
+
```

Comment:
What is the point of the batched function?  Are you assuming that the graph cannot fit into memory?